File: api-pytest-framework/features/steps/booking_patch_steps.py
"""Steps BDD pour booking_patch.feature -- PATCH /booking/{id}."""
from pytest_bdd import when, then
from pages.booking_page import BookingPage


# ── When : PATCH /booking/{id} ────────────────────────────────────────────────

@when('j\'envoie PATCH /booking/{id} avec {"firstname": "UpdatedName"}')
def patch_firstname(ctx):
    ctx["response"] = ctx["booking"].patch_booking(
        ctx["booking_id"], firstname="UpdatedName"
    )


@when('j\'envoie PATCH /booking/{id} avec {"totalprice": 999}')
def patch_totalprice(ctx):
    ctx["response"] = ctx["booking"].patch_booking(
        ctx["booking_id"], totalprice=999
    )


@when('j\'envoie PATCH /booking/{id} avec {"lastname": "Updated", "totalprice": 500}')
def patch_lastname_and_price(ctx):
    ctx["response"] = ctx["booking"].patch_booking(
        ctx["booking_id"], lastname="Updated", totalprice=500
    )


@when("j'envoie PATCH /booking/{id} sans header d'authentification")
def patch_no_token(ctx):
    ctx["response"] = BookingPage().patch_booking(ctx["booking_id"], firstname="X")


@when("j'envoie PATCH /booking/{id} avec un token invalide")
def patch_invalid_token(ctx):
    ctx["response"] = BookingPage(token="FAKE_TOKEN_INVALID").patch_booking(
        ctx["booking_id"], firstname="X"
    )


@when("j'envoie PATCH /booking/9999999 avec mon token")
def patch_inexistant(ctx):
    ctx["response"] = ctx["booking"].patch_booking(9999999, firstname="X")


@when("j'envoie PATCH /booking/{id} avec un body vide {}")
def patch_empty_body(ctx):
    ctx["response"] = ctx["booking"].patch_booking_empty(ctx["booking_id"])


# ── Then : assertions PATCH ───────────────────────────────────────────────────

@then("la reponse contient le nouveau firstname")
def check_new_firstname(ctx):
    body = ctx["response"].json()
    assert body.get("firstname") == "UpdatedName", (
        f"firstname non mis a jour : {body}"
    )


@then("la reponse contient totalprice = 999")
def check_new_totalprice(ctx):
    body = ctx["response"].json()
    assert body.get("totalprice") == 999, (
        f"totalprice non mis a jour : {body}"
    )


@then("la reponse contient le nouveau lastname et totalprice")
def check_lastname_and_price(ctx):
    body = ctx["response"].json()
    assert body.get("lastname")   == "Updated", f"lastname invalide : {body}"
    assert body.get("totalprice") == 500,       f"totalprice invalide : {body}"


@then("la reservation n'a pas ete modifiee")
def check_unchanged(ctx):
    # PATCH {} retourne 200 avec la réservation inchangée
    body = ctx["response"].json()
    assert "firstname" in body, f"Reponse inattendue apres PATCH vide : {body}"

File: api-Java-Rest-Assured/src/test/java/com/restfulbooker/steps/BookingPatchSteps.java
package com.restfulbooker.steps;

import com.restfulbooker.client.BookingClient;
import com.restfulbooker.context.ScenarioContext;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import io.restassured.path.json.JsonPath;

import static org.testng.Assert.assertEquals;
import static org.testng.Assert.assertNotNull;

/**
 * Steps BDD pour booking_patch.feature -- PATCH /booking/{id}.
 * Équivalent Java de features/steps/booking_patch_steps.py.
 */
public class BookingPatchSteps {

    private final ScenarioContext ctx;

    public BookingPatchSteps(ScenarioContext ctx) {
        this.ctx = ctx;
    }

    @When("^j'envoie PATCH /booking/\\{id\\} avec \\{\"firstname\": \"UpdatedName\"\\}$")
    public void patchFirstname() {
        ctx.setResponse(ctx.getBooking().patchBooking(ctx.getBookingId(), "UpdatedName", null, null));
    }

    @When("^j'envoie PATCH /booking/\\{id\\} avec \\{\"totalprice\": 999\\}$")
    public void patchTotalprice() {
        ctx.setResponse(ctx.getBooking().patchBooking(ctx.getBookingId(), null, null, 999));
    }

    @When("^j'envoie PATCH /booking/\\{id\\} avec \\{\"lastname\": \"Updated\", \"totalprice\": 500\\}$")
    public void patchLastnameAndPrice() {
        ctx.setResponse(ctx.getBooking().patchBooking(ctx.getBookingId(), null, "Updated", 500));
    }

    @When("^j'envoie PATCH /booking/\\{id\\} sans header d'authentification$")
    public void patchNoToken() {
        ctx.setResponse(new BookingClient().patchBooking(ctx.getBookingId(), "X", null, null));
    }

    @When("^j'envoie PATCH /booking/\\{id\\} avec un token invalide$")
    public void patchInvalidToken() {
        ctx.setResponse(new BookingClient("FAKE_TOKEN_INVALID").patchBooking(ctx.getBookingId(), "X", null, null));
    }

    @When("^j'envoie PATCH /booking/9999999 avec mon token$")
    public void patchInexistant() {
        ctx.setResponse(ctx.getBooking().patchBooking(9999999, "X", null, null));
    }

    @When("^j'envoie PATCH /booking/\\{id\\} avec un body vide \\{\\}$")
    public void patchEmptyBody() {
        ctx.setResponse(ctx.getBooking().patchBookingEmpty(ctx.getBookingId()));
    }

    // ── Then : assertions PATCH ───────────────────────────────────────────────

    @Then("^la reponse contient le nouveau lastname et totalprice$")
    public void checkLastnameAndPrice() {
        JsonPath json = ctx.getResponse().jsonPath();
        assertEquals(json.getString("lastname"), "Updated", "lastname invalide");
        assertEquals(json.getInt("totalprice"), 500, "totalprice invalide");
    }

    @Then("^la reservation n'a pas ete modifiee$")
    public void checkUnchanged() {
        JsonPath json = ctx.getResponse().jsonPath();
        assertNotNull(json.get("firstname"), "Reponse inattendue apres PATCH vide");
    }
}
